database/src/main/java/com/firebase/ui/database/IndexFirebaseArray.java

```diff
@@ -0,0 +1,123 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import android.support.v4.util.SimpleArrayMap;
+import android.util.Log;
+
+import com.google.firebase.database.DataSnapshot;
+import com.google.firebase.database.DatabaseError;
+import com.google.firebase.database.DatabaseReference;
+import com.google.firebase.database.Query;
+import com.google.firebase.database.ValueEventListener;
+
+import java.util.ArrayList;
+import java.util.List;
+
+class IndexFirebaseArray extends FirebaseArray {
+    private DatabaseReference mRef;
+    private OnChangedListener mListener;
+    private List<DataSnapshot> mSnapshots = new ArrayList<>();
+    private SimpleArrayMap<DatabaseReference, ValueEventListener> mValueEventListeners = new SimpleArrayMap<>();
+
+    public IndexFirebaseArray(Query keyRef, DatabaseReference dataRef) {
+        super(keyRef);
+        mRef = dataRef;
+    }
+
+    @Override
+    public void cleanup() {
+        super.cleanup();
+
+        for (int i = 0; i < mValueEventListeners.size(); i++) {
+            DatabaseReference ref = mValueEventListeners.keyAt(i);
+            ref.removeEventListener(mValueEventListeners.get(ref));
+        }
+    }
+
+    @Override
+    public int getCount() {
+        return mSnapshots.size();
+    }
+
+    @Override
+    public DataSnapshot getItem(int index) {
+        return mSnapshots.get(index);
+    }
+
+    @Override
+    public void setOnChangedListener(OnChangedListener listener) {
+        mListener = listener;
+    }
+
+    @Override
+    protected void notifyChangedListeners(OnChangedListener.EventType type, final int index, int oldIndex) {
+        if (mListener != null) {
+            switch (type) {
+                case Added:
+                    final boolean[] isNewListener = {true};
+
+                    ValueEventListener valueEventListener = new ValueEventListener() {
+                        @Override
+                        public void onDataChange(DataSnapshot dataSnapshot) {
+                            if (dataSnapshot.getValue() != null) {
+                                if (isNewListener[0]) {
+                                    mSnapshots.add(index, dataSnapshot);
+                                    mListener.onChanged(OnChangedListener.EventType.Added, index, -1);
+
+                                    isNewListener[0] = false;
+                                } else {
+                                    int index = getIndexForKey(dataSnapshot.getKey());
```

Comment:
Here I'm using the snapshots from `FirebaseArray`, not `IndexFirebaseArray`. I can't think of how this could be problematic since the keys are the same for both implementations, but I just wanted to put it out there anyway.

Also, this could be more performant since the snapshots from `FirebaseArray` should, in theory, be much smaller objects than the ones in `IndexFirebaseArray`.